Gracefully handle missing `pnpm`

`pnpm` is a required system dependency to create and work with Webstone apps. If `pnpm` is not available, we currently have a bad developer experience as seen below:

```
mike@a:~/ttt/webstone-tests$ npm create webstone-app@latest app

  ▄     ▄ ▄▄▄▄▄▄▄ ▄▄▄▄▄▄▄ ▄▄▄▄▄▄▄ ▄▄▄▄▄▄▄ ▄▄▄▄▄▄▄ ▄▄    ▄ ▄▄▄▄▄▄▄ 
  █ █ ▄ █ █       █  ▄    █       █       █       █  █  █ █       █
  █ ██ ██ █    ▄▄▄█ █▄█   █  ▄▄▄▄▄█▄     ▄█   ▄   █   █▄█ █    ▄▄▄█
  █       █   █▄▄▄█       █ █▄▄▄▄▄  █   █ █  █ █  █       █   █▄▄▄ 
  █       █    ▄▄▄█  ▄   ██▄▄▄▄▄  █ █   █ █  █▄█  █  ▄    █    ▄▄▄█
  █   ▄   █   █▄▄▄█ █▄█   █▄▄▄▄▄█ █ █   █ █       █ █ █   █   █▄▄▄ 
  █▄▄█ █▄▄█▄▄▄▄▄▄▄█▄▄▄▄▄▄▄█▄▄▄▄▄▄▄█ █▄▄▄█ █▄▄▄▄▄▄▄█▄█  █▄▄█▄▄▄▄▄▄▄█
  
  
✔ Creating the application directory
✔ Creating the application directory
  ✔ Determining app directory name
  ✔ Creating app directory
✔ Preparing the project structure
  ✔ Copying the template
❯ Installing the application foundation
  ✔ Initializing the web service
  ✖ Command failed with exit code 127: pnpm install
    /bin/sh: 1: pnpm: not found
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: Command failed with exit code 127: pnpm install
/bin/sh: 1: pnpm: not found
    at makeError (file:///home/mike/.npm/_npx/7f9158b99959dcd2/node_modules/execa/lib/error.js:59:11)
    at handlePromise (file:///home/mike/.npm/_npx/7f9158b99959dcd2/node_modules/execa/index.js:119:26)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Task.run (file:///home/mike/.npm/_npx/7f9158b99959dcd2/node_modules/listr2/dist/index.js:960:11)
    at async /home/mike/.npm/_npx/7f9158b99959dcd2/node_modules/p-map/index.js:57:22 {
  shortMessage: 'Command failed with exit code 127: pnpm install',
  command: 'pnpm install',
  escapedCommand: 'pnpm install',
  exitCode: 127,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: '/bin/sh: 1: pnpm: not found',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
npm ERR! code 1
npm ERR! path /home/mike/ttt/webstone-tests
npm ERR! command failed
npm ERR! command sh -c create-webstone-app "app"

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/mike/.npm/_logs/2022-07-18T02_14_33_869Z-debug-0.log
mike@a:~/ttt/webstone-tests$
```

We should fix this with a `Listr` task to validate system requirements. While our goal is to keep system dependencies to a minimum, there may be more down the road and we can extend the Listr system requirements check task accordingly.

--- a/.changeset/eleven-terms-double.md
+++ b/.changeset/eleven-terms-double.md
@@ -1,0 +1,5 @@
+---
+"create-webstone-app": patch
+---
+
+Run a system requirements check before creating a Webstone app.

--- a/packages/create-webstone-app/src/helpers.ts
+++ b/packages/create-webstone-app/src/helpers.ts
@@ -1,16 +1,6 @@
-import { execa } from "execa";
-import fs from "fs-extra";
-import path from "path";
 import chalk from "chalk";
-//@ts-ignore this package doesn't provdide a declaration file
-import { create } from "create-svelte";
-import { dirname } from "path";
-import { fileURLToPath } from "url";
 
-import { ListrTaskWrapper, ListrRenderer, ListrTask } from "listr2/dist/index";
-import { Listr } from "listr2";
-
-const __dirname = dirname(fileURLToPath(import.meta.url));
+import { ListrTaskWrapper, ListrRenderer } from "listr2/dist/index";
 
 export interface Ctx {
   appDir: string;
@@ -35,76 +25,6 @@ export const displayWelcome = () =>
     resolve();
   });
 
-const determineAppDirName = async (ctx: Ctx, task: WebstoneTask) => {
-  const appName = process.argv[2];
-  ctx.appDir = appName ? appName.toLowerCase().replace(/\s/g, "-") : ".";
-
-  task.output = `App directory name: ${ctx.appDir}`;
-  task.output = ctx.appDir;
-  return;
-};
-
-export const createAppDir = async (ctx: Ctx, task: WebstoneTask) => {
-  const appDir = ctx.appDir;
-
-  if (fs.existsSync(appDir)) {
-    if (fs.readdirSync(appDir).length > 0) {
-      const response = await task.prompt({
-        type: "confirm",
-        name: "value",
-        message: `The ./${appDir} directory is not empty. Do you want to overwrite it?`,
-        initial: false,
-      });
-
-      if (!response.value) {
-        throw new Error(
-          `Exiting, please empty the ./${appDir} directory or choose a different one to create the Webstone app.`
-        );
-      }
-
-      await fs.rm(appDir, { recursive: true, force: true });
-    }
-  }
-  fs.mkdirSync(appDir, { recursive: true });
-  return appDir;
-};
-
-export const copyTemplate = (ctx: Ctx) => {
-  const templateDir = path.join(__dirname, "..", "template");
-  fs.copySync(templateDir, ctx.appDir);
-};
-
-const initWebApp = async (ctx: Ctx) => {
-  const webAppDir = `${ctx.appDir}/services/web`;
-  ctx.webAppDir = webAppDir;
-  console.log(`Installing web app in ${webAppDir}...`);
-
-  try {
-    fs.removeSync(`${webAppDir}/.keep`);
-
-    await create(webAppDir, {
-      name: "webstone-app",
-      template: "skeleton",
-      types: "typescript",
-      prettier: true,
-      eslint: true,
-      playwright: true,
-    });
-  } catch (error) {
-    console.error(error);
-    process.exit(1);
-  }
-};
-
-const installWebAppDependencies = async (ctx: Ctx) => {
-  const installProcess = execa("pnpm", ["install"], {
-    cwd: ctx.webAppDir,
-    shell: true,
-  });
-
-  return installProcess;
-};
-
 export const displayNextSteps = async (ctx: Ctx) => {
   console.log(`
 ===================================================
@@ -121,62 +41,3 @@ Next steps:
   - ${chalk.bold(chalk.cyan("pnpm ws dev"))}
     `);
 };
-
-const createAppDirTasks: ListrTask[] = [
-  {
-    task: determineAppDirName,
-    title: "Determining app directory name",
-  },
-  {
-    task: createAppDir,
-    title: "Creating app directory",
-  },
-];
-
-const createCopyTemplateTasks: ListrTask[] = [
-  {
-    task: copyTemplate,
-    title: "Copying the template",
-    options: { persistentOutput: true },
-  },
-];
-
-const createInstallWebAppTasks: ListrTask[] = [
-  {
-    task: initWebApp,
-    title: "Initializing the web service",
-  },
-  {
-    task: installWebAppDependencies,
-    title: "Installing dependencies for the web service",
-    options: { bottomBar: true },
-  },
-];
-
-export const tasks = new Listr<Ctx>(
-  [
-    {
-      task(_, task) {
-        return task.newListr(createAppDirTasks);
-      },
-      title: "Creating the application directory",
-    },
-    {
-      task(_, task) {
-        return task.newListr(createCopyTemplateTasks);
-      },
-      title: "Preparing the project structure",
-    },
-    {
-      task(_, task) {
-        return task.newListr(createInstallWebAppTasks);
-      },
-      title: "Installing the application foundation",
-    },
-  ],
-  {
-    rendererOptions: {
-      collapse: false,
-    },
-  }
-);

--- a/packages/create-webstone-app/src/index.ts
+++ b/packages/create-webstone-app/src/index.ts
@@ -1,6 +1,5 @@
-// FIXME: Does anyone know how to properly type the `pipe` function?
-/* eslint @typescript-eslint/no-explicit-any: "off" */
-import { tasks, displayWelcome, displayNextSteps } from "./helpers.js";
+import { displayWelcome, displayNextSteps } from "./helpers.js";
+import { tasks } from "./tasks/index.js";
 
 displayWelcome();
 const context = await tasks.run();

--- a/packages/create-webstone-app/src/tasks/0-check-system-requirements/index.ts
+++ b/packages/create-webstone-app/src/tasks/0-check-system-requirements/index.ts
@@ -1,0 +1,27 @@
+import { execa } from "execa";
+import { ListrTask } from "listr2/dist/index";
+
+const checkPnpm = async () => {
+  try {
+    await execa("pnpm", ["version"]);
+    return;
+  } catch (error) {
+    throw new Error("Please install `pnpm` globally with `npm i -g pnpm`");
+  }
+};
+
+const checkSystemRequirements: ListrTask[] = [
+  {
+    task: checkPnpm,
+    title: "Validating pnpm...",
+  },
+];
+
+export const tasks: ListrTask[] = [
+  {
+    task(_, task) {
+      return task.newListr(checkSystemRequirements);
+    },
+    title: "Checking system requirements",
+  },
+];

--- a/packages/create-webstone-app/src/tasks/1-create-app-directory/index.ts
+++ b/packages/create-webstone-app/src/tasks/1-create-app-directory/index.ts
@@ -1,0 +1,58 @@
+import fs from "fs-extra";
+import { ListrTask } from "listr2/dist/index";
+
+import { Ctx, WebstoneTask } from "../../helpers";
+
+const determineAppDirName = async (ctx: Ctx, task: WebstoneTask) => {
+  const appName = process.argv[2];
+  ctx.appDir = appName ? appName.toLowerCase().replace(/\s/g, "-") : ".";
+
+  task.output = `App directory name: ${ctx.appDir}`;
+  task.output = ctx.appDir;
+  return;
+};
+
+const createAppDir = async (ctx: Ctx, task: WebstoneTask) => {
+  const appDir = ctx.appDir;
+
+  if (fs.existsSync(appDir)) {
+    if (fs.readdirSync(appDir).length > 0) {
+      const response = await task.prompt({
+        type: "confirm",
+        name: "value",
+        message: `The ./${appDir} directory is not empty. Do you want to overwrite it?`,
+        initial: false,
+      });
+
+      if (!response.value) {
+        throw new Error(
+          `Exiting, please empty the ./${appDir} directory or choose a different one to create the Webstone app.`
+        );
+      }
+
+      await fs.rm(appDir, { recursive: true, force: true });
+    }
+  }
+  fs.mkdirSync(appDir, { recursive: true });
+  return appDir;
+};
+
+const createAppDirTasks: ListrTask[] = [
+  {
+    task: determineAppDirName,
+    title: "Determining app directory name",
+  },
+  {
+    task: createAppDir,
+    title: "Creating app directory",
+  },
+];
+
+export const tasks: ListrTask[] = [
+  {
+    task(_, task) {
+      return task.newListr(createAppDirTasks);
+    },
+    title: "Creating the application directory",
+  },
+];

--- a/packages/create-webstone-app/src/tasks/1-create-app-directory/index.ts
+++ b/packages/create-webstone-app/src/tasks/1-create-app-directory/index.ts
@@ -12,7 +12,7 @@ const determineAppDirName = async (ctx: Ctx, task: WebstoneTask) => {
   return;
 };
 
-const createAppDir = async (ctx: Ctx, task: WebstoneTask) => {
+export const createAppDir = async (ctx: Ctx, task: WebstoneTask) => {
   const appDir = ctx.appDir;
 
   if (fs.existsSync(appDir)) {

--- a/packages/create-webstone-app/src/tasks/2-prepare-project-structure/index.ts
+++ b/packages/create-webstone-app/src/tasks/2-prepare-project-structure/index.ts
@@ -1,0 +1,30 @@
+import fs from "fs-extra";
+import path, { dirname } from "path";
+import { ListrTask } from "listr2/dist/index";
+import { fileURLToPath } from "url";
+
+import { Ctx } from "../../helpers";
+
+const __dirname = dirname(fileURLToPath(import.meta.url));
+
+export const copyTemplate = (ctx: Ctx) => {
+  const templateDir = path.join(__dirname, "../../..", "template");
+  fs.copySync(templateDir, ctx.appDir);
+};
+
+const createCopyTemplateTasks: ListrTask[] = [
+  {
+    task: copyTemplate,
+    title: "Copying the template",
+    options: { persistentOutput: true },
+  },
+];
+
+export const tasks: ListrTask[] = [
+  {
+    task(_, task) {
+      return task.newListr(createCopyTemplateTasks);
+    },
+    title: "Preparing the project structure",
+  },
+];

--- a/packages/create-webstone-app/src/tasks/3-install-app-dependencies/index.ts
+++ b/packages/create-webstone-app/src/tasks/3-install-app-dependencies/index.ts
@@ -1,0 +1,59 @@
+//@ts-ignore this package doesn't provdide a declaration file
+import { create } from "create-svelte";
+import { execa } from "execa";
+import fs from "fs-extra";
+import { ListrTask } from "listr2/dist/index";
+
+import { Ctx } from "../../helpers";
+
+const initWebApp = async (ctx: Ctx) => {
+  const webAppDir = `${ctx.appDir}/services/web`;
+  ctx.webAppDir = webAppDir;
+  console.log(`Installing web app in ${webAppDir}...`);
+
+  try {
+    fs.removeSync(`${webAppDir}/.keep`);
+
+    await create(webAppDir, {
+      name: "webstone-app",
+      template: "skeleton",
+      types: "typescript",
+      prettier: true,
+      eslint: true,
+      playwright: true,
+    });
+  } catch (error) {
+    console.error(error);
+    process.exit(1);
+  }
+};
+
+const installWebAppDependencies = async (ctx: Ctx) => {
+  const installProcess = execa("pnpm", ["install"], {
+    cwd: ctx.webAppDir,
+    shell: true,
+  });
+
+  return installProcess;
+};
+
+const createInstallWebAppTasks: ListrTask[] = [
+  {
+    task: initWebApp,
+    title: "Initializing the web service",
+  },
+  {
+    task: installWebAppDependencies,
+    title: "Installing dependencies for the web service",
+    options: { bottomBar: true },
+  },
+];
+
+export const tasks: ListrTask[] = [
+  {
+    task(_, task) {
+      return task.newListr(createInstallWebAppTasks);
+    },
+    title: "Installing the application foundation",
+  },
+];

--- a/packages/create-webstone-app/src/tasks/index.ts
+++ b/packages/create-webstone-app/src/tasks/index.ts
@@ -1,0 +1,21 @@
+import { Listr } from "listr2";
+
+import { Ctx } from "../helpers.js";
+import { tasks as checkSystemRequirements } from "./0-check-system-requirements/index.js";
+import { tasks as createAppDirectoryTasks } from "./1-create-app-directory/index.js";
+import { tasks as prepareProjectStructureTasks } from "./2-prepare-project-structure/index.js";
+import { tasks as installAppDependenciesTasks } from "./3-install-app-dependencies/index.js";
+
+export const tasks = new Listr<Ctx>(
+  [
+    ...checkSystemRequirements,
+    ...createAppDirectoryTasks,
+    ...prepareProjectStructureTasks,
+    ...installAppDependenciesTasks,
+  ],
+  {
+    rendererOptions: {
+      collapse: false,
+    },
+  }
+);

--- a/packages/create-webstone-app/tests/helpers/copy-template.ts
+++ b/packages/create-webstone-app/tests/helpers/copy-template.ts
@@ -3,7 +3,8 @@ import path from "path";
 import sinon from "sinon";
 import { test } from "uvu";
 import * as assert from "uvu/assert";
-import { copyTemplate, Ctx } from "../../src/helpers";
+import { Ctx } from "../../src/helpers";
+import { copyTemplate } from "../../src/tasks/2-prepare-project-structure";
 
 const context: Ctx = {
   appDir: "test-dir",

--- a/packages/create-webstone-app/tests/helpers/create-app-dir.ts
+++ b/packages/create-webstone-app/tests/helpers/create-app-dir.ts
@@ -2,7 +2,8 @@ import fs from "fs-extra";
 import sinon from "sinon";
 import { test } from "uvu";
 import * as assert from "uvu/assert";
-import { createAppDir, Ctx, WebstoneTask } from "../../src/helpers";
+import { Ctx, WebstoneTask } from "../../src/helpers";
+import { createAppDir } from "../../src/tasks/1-create-app-directory";
 
 test.before.each(() => {
   sinon.replace(console, "log", sinon.fake());

